### DISCUSSION TITLE
[Joke] Port this project to Rust

### DISCUSSION THREAD

**Discussion by stevefan1999-personal** [hasura/graphql-engine#6093] [General] [answered]
I know it is just a joke haha. This is such a cosmic idea: who the f**k would rewrite something that already works? But you can see that we actually have someone brave enough to write a Haskell-Rust transpiler [tcr/corollary](https://github.com/tcr/corollary) that still doesn't work. This however sparked some hope to in some day let us go away from Haskell.

Nonetheless, Rust ecosystem/Cargo is still very early on and fragile, that not even Tokio is stable yet (but going to next month). Compare that to the ecosystem Haskell had, who had more than 25 years of history aging along with GHC, is comparably more stable than Rust and thus less prone to f**k up. 

The problem with Haskell is that it is way too hard to learn, that you need to know category theory, type theory and to some degree abstract algebra (which hurts my brain) for you to fully understand how Haskell works and how to write idiomatic Haskell code. All those endomorphism/automorphism/monomorphism sounds like alien languages to me even with visual aid. WTF is a subgroup/monad/semiring? I feel like a complete idiot trying to know Haskell, let alone trying to learn it, understand it.

But also combining it with what I said about Haskell is less likely to f**k up, this is one of the evidence why and why it is also an advantage of Haskell: the people who can understand Haskell are more likely to have higher mathematical intelligent (or maybe high-functioning autists who can battle on IMO/ACM), therefore the program they produce are usually more likely to be correct, flawless, cleaner architecture, more useful and will work properly over the decay of longer time. This also means people who wrote Haskell are also elites among elites.

While Rust -- although still in the category of "hard" languages for most of us, and should belong to the same high-ceiling with Haskell (but Idris or Coq they probably sit on the very tip), is ten-fold easier than Haskell from my perspective, in my learning experience. 

You would have to trade some Haskell gucci such as:
1. HKT (The way you do it in Rust is actually to emulate it using something called "URI", derived from this [paper](https://www.cl.cam.ac.uk/~jdy22/papers/lightweight-higher-kinded-polymorphism.pdf))
2. Currying
3. HM type system (for true automatic type deduction)
4. QuickCheck, although there is a Rust partial equivalent, but not as powerful as it should had been.
5. Automatic garbage collector (this is disputable, some stubborn gurus view GC as bane, most people incl. me actually see it as a boon)

But in return you would have more people to contribute to this project because of, well, the lower ceiling. And you can write less code because Cargo is _expanding_. You can find many gems like Hyper, Diesel, Actix...etc to cover your needs.

I'm already sensing the wave of rage from the people who can "understand" Haskell, to bomb on this thread. You have the right to be such arrogant though because you have the right to look down on newbies.

**Comment by tylerhannan** [comment, marked as answer]
Thanks for the perspective...

For now we are happy with our chosen language of implementation. If we choose to reconsider/reimplement in the future than we would also consider other languages.
